strutil/shlex/shlex.go

```diff
@@ -414,3 +414,30 @@ func Split(s string) ([]string, error) {
 		subStrings = append(subStrings, word)
 	}
 }
+
+// Join returns a shell-escaped string version of the arguments in s.
+func Join(s []string) string {
+	quoted := make([]string, len(s))
+	for i, x := range s {
+		quoted[i] = Quote(x)
+	}
+	return strings.Join(quoted, " ")
+}
+
+// Quote returns a shell-escaped version of the single argument s.
+func Quote(s string) string {
+	if s == "" {
+		return `''`
+	}
+	for _, c := range s {
+		switch {
+		case c >= 'a' && c <= 'z':
+		case c >= 'A' && c <= 'Z':
+		case c >= '0' && c <= '9':
+		case strings.ContainsRune("_@%+=:,./-", c):
+		default:
+			return "'" + strings.ReplaceAll(s, "'", "'\"'\"'") + "'"
```

Comment:
For whatever reason it took me a while to figure out that these were the safe characters. Maybe add a small comment like so?

```go
default:
    // The above cases are all the safe characters. If there are any other characters,
    // quote the entire input.
    return "'" + strings.ReplaceAll(s, "'", "'\"'\"'") + "'"
```

Reply:
Thanks for the wording Ben. Updated in https://github.com/canonical/x-go/pull/19/commits/08c9987a9a6f53281d401846358666dcb4c07a12.